cpp/order/Steinhardt.cc

```diff
@@ -184,80 +236,117 @@ void Steinhardt::computeAve(const freud::locality::NeighborList* nlist,
+        for (size_t k = 0; k < m_num_ms[l_index]; ++k)
+        {
+            // Add the norm, which is the complex squared magnitude
+            calc_norm += norm(qlm[k]);
+        }
+        const float ql_system_norm = std::sqrt(calc_norm * normalizationfactor);
 
-    if (m_wl)
-    {
-        auto wigner3jvalues = getWigner3j(m_l);
-        float wl_system_norm = reduceWigner3j(m_qlm.get(), m_l, wigner3jvalues);
+        if (m_wl)
+        {
+            const auto wigner3j_values = getWigner3j(l);
+            float wl_system_norm = reduceWigner3j(qlm.get(), l, wigner3j_values);
 
-        // The normalization factor of wl is calculated using qli, which is
-        // equivalent to calculate the normalization factor from qlmi
-        if (m_wl_normalize)
+            // The normalization factor of wl is calculated using qli, which is
+            // equivalent to calculate the normalization factor from qlmi
+            if (m_wl_normalize)
+            {
+                const float wl_normalization = std::sqrt(normalizationfactor) / ql_system_norm;
+                wl_system_norm *= wl_normalization * wl_normalization * wl_normalization;
+            }
+            system_norms[l_index] = wl_system_norm;
+        }
+        else
         {
-            const float wl_normalization = std::sqrt(normalizationfactor) / ql_system_norm;
-            wl_system_norm *= wl_normalization * wl_normalization * wl_normalization;
+            system_norms[l_index] = ql_system_norm;
         }
-        return wl_system_norm;
     }
-
-    return ql_system_norm;
+    return system_norms;
 }
 
 void Steinhardt::aggregatewl(util::ManagedArray<float>& target,
-                             const util::ManagedArray<std::complex<float>>& source,
+                             const std::vector<util::ManagedArray<std::complex<float>>>& source,
                              const util::ManagedArray<float>& normalization_source) const
 {
-    auto wigner3jvalues = getWigner3j(m_l);
-    const auto normalizationfactor = float(4.0 * M_PI / m_num_ms);
     util::forLoopWrapper(0, m_Np, [&](size_t begin, size_t end) {
         for (size_t i = begin; i < end; ++i)
         {
-            target[i] = reduceWigner3j(&(source({static_cast<unsigned int>(i), 0})), m_l, wigner3jvalues);
-            if (m_wl_normalize)
+            const auto target_particle_index = target.getIndex({i, 0});
+            const auto norm_particle_index = normalization_source.getIndex({i, 0});
+            for (size_t l_index = 0; l_index < m_ls.size(); ++l_index)
             {
-                const float normalization = std::sqrt(normalizationfactor) / normalization_source[i];
-                target[i] *= normalization * normalization * normalization;
+                const auto l = m_ls[l_index];
+                const auto& source_l = source[l_index];
+
+                const auto normalizationfactor = static_cast<float>(4.0 * M_PI / m_num_ms[l_index]);
```

Comment:
Considering how often we compute this normalization factor, perhaps it would be worth defining a compile-time evaluated template function `getNormalizationFactor<num_ms>()` encoding the values for the supported range of `l` values? Currently we're limited anyway by the values of the Wigner3j coefficients we hard code (I plan to replace that eventually), so you could match that with a compile-time generation of the normalization. I'm happy to take that on myself at a later point if you want to punt, though.

Reply:
Punting